offchainservice/src/main/java/pl/edu/agh/blockchain/offchainservice/service/RegistrationService.java

```diff
@@ -0,0 +1,113 @@
+import pl.edu.agh.blockchain.offchainservice.exceptions.*;
+import pl.edu.agh.blockchain.offchainservice.model.PendingVerification;
+import pl.edu.agh.blockchain.offchainservice.model.User;
+import pl.edu.agh.blockchain.offchainservice.repository.EmailRepository;
+import pl.edu.agh.blockchain.offchainservice.repository.PendingRepository;
+import pl.edu.agh.blockchain.offchainservice.utils.CodeGenerator;
+
+import java.io.IOException;
+import java.text.MessageFormat;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class RegistrationService {
+
+    private final EmailRepository emailRepository;
+
+    private final PendingRepository pendingRepository;
+
+    private final EMailService eMailService;
+
+    private final DeploymentService deploymentService;
+
+    private final ModelMapper modelMapper = new ModelMapper();
+
+    public RegistrationService(EmailRepository emailRepository, EMailService eMailService,
+                               PendingRepository pendingRepository, DeploymentService deploymentService) {
+        this.emailRepository = emailRepository;
+        this.eMailService = eMailService;
+        this.pendingRepository = pendingRepository;
+        this.deploymentService = deploymentService;
+    }
+
+    public void verifyUser(String email) {
+        if (!email.endsWith("@agh.edu.pl") && !email.endsWith("@student.agh.edu.pl")) {
+            throw new InvalidEmailDomainException(email + " is not AGH email");
+        }
+        if (!emailRepository.existsByEmail(email)) {
+            try {
+                String token = CodeGenerator.generateNewToken();
+                eMailService.sendEmail(email, token);
+                savePendingVerification(email, token);
+            } catch (IOException e) {
+                throw new SendFailedException(e.getMessage());
+            }
+        } else {
+            User user = emailRepository.findByEmail(email).get(0);
+            throw new UserAlreadyExistsException(MessageFormat.format("User with mail: {0} was registered: {1}", user.getEmail(),
+                    user.getCreationDate()));
+        }
+    }
+
+    public void registerUser(RegisterInformationDTO registerInformationDTO) {
+        Optional<PendingVerificationDTO> pendingVerificationDTO = registrationPending(registerInformationDTO.getEmail());
+        if (pendingVerificationDTO.isPresent()) {
+            PendingVerificationDTO pendingVerification = pendingVerificationDTO.get();
+            if (!correctToken(registerInformationDTO, pendingVerification)) {
+                throw new InvalidVerificationToken("Token is not correct. Check your mail once again or wait about 20 minutes and generate new token");
+            }
+            saveUser(registerInformationDTO.getEmail());
+            removeFromPending(registerInformationDTO.getEmail());
+            deploymentService.deployContract(registerInformationDTO.getBlockchainAddress());
```

Comment:
The naming is a bit misleading here. We don't need to 'deploy' a contract at this moment. We need to call `addStudentAddress(registerInformationDTO.getBlockchainAddress())` or `addEmployeeAddress(registerInformationDTO.getBlockchainAddress())` on the deployed instance of the `VotingFactory` contract (method depends on the email ending).